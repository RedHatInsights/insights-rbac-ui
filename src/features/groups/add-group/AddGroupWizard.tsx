import React, { useMemo, useRef, useState } from 'react';
import { useDispatch } from 'react-redux';
import useChrome from '@redhat-cloud-services/frontend-components/useChrome';
import { useFlag } from '@unleash/proxy-client-react';
import { addNotification } from '@redhat-cloud-services/frontend-components-notifications/';
import FormRenderer from '@data-driven-forms/react-form-renderer/form-renderer';
import Pf4FormTemplate from '@data-driven-forms/pf4-component-mapper/form-template';
import componentMapper from '@data-driven-forms/pf4-component-mapper/component-mapper';
import WarningModal from '@patternfly/react-component-groups/dist/dynamic/WarningModal';
import { schemaBuilder } from './schema';
import { addGroup, addServiceAccountsToGroup } from '../../../redux/groups/actions';
import { SetName } from './components/stepName/SetName';
import { SetRoles } from './components/stepRoles/SetRoles';
import { SetUsers } from './components/stepUsers/SetUsers';
import SetServiceAccounts from './components/stepServiceAccounts/SetServiceAccounts';
import { SummaryContent } from './components/stepReview/SummaryContent';
import { AddGroupSuccess } from './AddGroupSuccess';
import useAppNavigate from '../../../hooks/useAppNavigate';
import paths from '../../../utilities/pathnames';
import { AddGroupWizardContext } from './add-group-wizard-context';

const FormTemplate = (props: React.ComponentProps<typeof Pf4FormTemplate>) => <Pf4FormTemplate {...props} showFormControls={false} />;

// Component mapper extension for wizard steps
const mapperExtension = {
  'set-name': SetName,
  'set-roles': SetRoles,
  'set-users': SetUsers,
  'set-service-accounts': SetServiceAccounts,
  'summary-content': SummaryContent,
  'add-group-success': AddGroupSuccess,
};

interface AddGroupWizardProps {
  // No direct props - component receives data from URL params and context
}

export const AddGroupWizard: React.FC<AddGroupWizardProps> = () => {
  const dispatch = useDispatch();
  const navigate = useAppNavigate();
  const chrome = useChrome();

  // Feature flags for wizard configuration
  const enableServiceAccounts =
    (chrome.isBeta() && useFlag('platform.rbac.group-service-accounts')) ||
    (!chrome.isBeta() && useFlag('platform.rbac.group-service-accounts.stable'));

  const enableWorkspaces = useFlag('platform.rbac.workspaces');

  // When workspaces is enabled, roles are disabled (per business logic)
  const enableRoles = !enableWorkspaces;

  const [cancelWarningVisible, setCancelWarningVisible] = useState<boolean>(false);
  const [submitting] = useState<boolean>(false);
  const [submittingError] = useState<string | null>(null);
  const [wizardError, setWizardError] = useState<boolean | undefined>(undefined);
  const [wizardContextValue, setWizardContextValue] = useState<{
    success: boolean;
    submitting: boolean;
    error: string | null;
    setHideForm: () => void;
    setWizardSuccess: (success: boolean) => void;
  }>({
    success: false,
    submitting: false,
    error: null,
    setHideForm: () => {},
    setWizardSuccess: () => {},
  });

  const container = useRef<HTMLDivElement>(null);

  interface FormData {
    'group-name': string;
    'group-description': string;
    'users-list'?: Array<{ label: string; username?: string; value?: string }>;
    'roles-list'?: Array<{ uuid: string; value?: string }>;
    'set-service-accounts'?: string[];
  }

  const onSubmit = async (formData: FormData) => {
    const {
      'group-name': name,
      'group-description': description,
      'users-list': users,
      'roles-list': roles,
      'set-service-accounts': serviceAccounts,
    } = formData;

    // Prepare all group data for single API call
    const groupData = {
      name,
      description,
      // Add users if any are selected
      user_list: users && users.length > 0 ? users.map((user) => ({ username: user.username || user.label || user.value || '' })) : [],
      // Add roles if any are selected
      roles_list: roles && roles.length > 0 ? roles.map((role) => role.uuid || role.value || '').filter(Boolean) : [],
    };

    try {
      // Create group with all data in one call
      const newGroupAction = dispatch(addGroup(groupData)) as { payload: Promise<any> };
      const newGroup = await newGroupAction.payload;

      // Check if group creation actually succeeded
      if (newGroup?.error) {
        throw new Error('Group creation returned error: ' + JSON.stringify(newGroup.error));
      }

      // Handle service accounts separately (if enabled)
      if (serviceAccounts && serviceAccounts.length > 0) {
        const serviceAccountObjects = serviceAccounts.map((sa) => ({ uuid: sa }));
        const serviceAccountAction = dispatch(addServiceAccountsToGroup(newGroup.uuid, serviceAccountObjects)) as { payload: Promise<unknown> };
        await serviceAccountAction.payload;
      }

      // Success! Show notification
      dispatch(
        addNotification({
          variant: 'success',
          title: 'Group created successfully',
          description: 'The group has been created and configured successfully.',
        }),
      );

      // Navigate to the new group or groups list (in separate try-catch to not fail group creation)
      try {
        if (newGroup?.uuid) {
          const pathname = `${paths['group-detail'].link.split('/').slice(0, -1).join('/')}/${newGroup.uuid}`;
          navigate(pathname);
        } else {
          console.warn('⚠️ No UUID returned, navigating to groups list');
          navigate('/groups');
        }
      } catch (navError) {
        console.error('❌ Navigation error (but group was created successfully):', navError);
        // Still try to go back to groups list
        navigate('/groups');
      }
    } catch (error) {
      console.error('❌ Error creating group:', error);
      dispatch(
        addNotification({
          variant: 'danger',
          title: 'Error creating group',
          description: 'There was an error creating the group. Please try again.',
        }),
      );
    } finally {
    }
  };

  const onCancelWarningConfirm = () => {
    setCancelWarningVisible(false);
    navigate('/groups');
  };

  const schema = schemaBuilder(container.current, enableServiceAccounts, enableRoles);

  const contextValue = useMemo(
    () => ({
      ...wizardContextValue,
      success: false,
      submitting,
      submittingGroup: false,
      submittingServiceAccounts: false,
      error: wizardError !== undefined ? wizardError : submittingError || undefined,
      setHideForm: () => {},
      setWizardSuccess: (success: boolean) => setWizardContextValue((prev) => ({ ...prev, success })),
      setWizardError,
    }),
    [wizardContextValue, wizardError, submitting, submittingError, setWizardError],
  );

  return (
    <AddGroupWizardContext.Provider value={contextValue}>
      <div ref={container}>
        {cancelWarningVisible && (
          <WarningModal
            isOpen={cancelWarningVisible}
            title="Exit group creation?"
            confirmButtonLabel="Exit"
            onClose={() => setCancelWarningVisible(false)}
            onConfirm={onCancelWarningConfirm}
          >
            {'All unsaved changes will be lost. Are you sure you want to exit?'}
          </WarningModal>
        )}
        <FormRenderer
          schema={schema}
          subscription={{ values: true }}
          FormTemplate={FormTemplate}
          componentMapper={{ ...componentMapper, ...mapperExtension }}
          onSubmit={onSubmit}
          onCancel={() => setCancelWarningVisible(true)}
        />
      </div>
    </AddGroupWizardContext.Provider>
  );
};

export default AddGroupWizard;
