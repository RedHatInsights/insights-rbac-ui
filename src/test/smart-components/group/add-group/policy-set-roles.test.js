import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import thunk from 'redux-thunk';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { MemoryRouter, Route } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import AddGroupWizard from '../../../../smart-components/group/add-group/add-group-wizard';

describe('<SetRoles />', () => {
  let initialProps;
  let initialState;
  const middlewares = [thunk, promiseMiddleware, notificationsMiddleware()];
  let mockStore;

  const GroupWrapper = ({ store, children }) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={['/policies/', '/policy/add-role/']} initialIndex={1} keyLength={0}>
        {children}
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {
      selectedRoles: [],
      setSelectedRoles: jest.fn(),
      roles: [],
      title: 'Test set roles',
      description: 'Test for set roles page',
    };
    initialState = {
      roleReducer: {
        roles: [
          {
            uuid: '123',
            value: 'foo',
          },
        ],
      },
      groupReducer: {
        groups: {
          data: [
            {
              uuid: '123',
              name: 'SampleGroup',
            },
          ],
        },
      },
    };
    mockStore = configureStore(middlewares);
  });

  it('should render correctly', () => {
    Date.now = jest.fn(() => '123');
    const store = mockStore(initialState);
    const wrapper = mount(
      <GroupWrapper store={store}>
        <Route path="/groups/add-group/" render={() => <AddGroupWizard {...initialProps} />} />
      </GroupWrapper>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
