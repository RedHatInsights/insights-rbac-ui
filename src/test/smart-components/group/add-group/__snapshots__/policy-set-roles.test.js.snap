// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SetRoles /> should render correctly 1`] = `
<GroupWrapper
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <MemoryRouter
      initialEntries={
        Array [
          "/policies/",
          "/policy/add-role/",
        ]
      }
      initialIndex={1}
      keyLength={0}
    >
      <Router
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "pathname": "/policies/",
                "search": "",
                "state": undefined,
              },
              Object {
                "hash": "",
                "pathname": "/policy/add-role/",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 1,
            "length": 2,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/policy/add-role/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <Route
          path="/groups/add-group/"
          render={[Function]}
        />
      </Router>
    </MemoryRouter>
  </Provider>
</GroupWrapper>
`;
